File: Medium/1277. Count Square Submatrices with All Ones.py
from typing import List


class DP:
    def countSquares(self, matrix: List[List[int]]) -> int:
        cnt = sum(matrix[i][0] for i in range(len(matrix))) + sum(matrix[0][i] for i in range(1, len(matrix[0])))
        for r in range(1, len(matrix)):
            for c in range(1, len(matrix[0])):
                if not matrix[r][c]:
                    continue
                matrix[r][c] += min(matrix[r - 1][c], matrix[r][c - 1], matrix[r - 1][c - 1])
                cnt += matrix[r][c]
        print(matrix)
        return cnt

class PrefixSum:
    def countSquares(self, matrix: List[List[int]]) -> int:
        m, n = len(matrix), len(matrix[0])
        cnt = sum(1 if matrix[i][0] else 0 for i in range(m))
        for r in range(m):
            for c in range(1, n):
                if matrix[r][c] == 0:
                    continue
                cnt += 1
                matrix[r][c] += matrix[r][c - 1]
                curr_min = matrix[r][c]
                for s in range(1, r + 1):
                    curr_min = min(curr_min, matrix[r - s][c])
                    if curr_min <= s:
                        break
                    cnt += 1
        return cnt

File: Medium/1277. Count Square Submatrices with All Ones.java
class Solution {
    public int countSquares(int[][] matrix) {
        int R = matrix.length, C = matrix[0].length, result = 0;
        for (int r = 0; r < R; r++) {
            for (int c = 0; c < C; c++) {
                if (matrix[r][c] == 1 && r > 0 && c > 0) {
                    matrix[r][c] = Math.min(matrix[r - 1][c - 1], Math.min(matrix[r - 1][c], matrix[r][c - 1])) + 1;
                }
                result += matrix[r][c];
            }
        }
        return result;
    }
}